File: VB/HeadFirstPatterns/DoFactory.HeadFirst/DoFactory.HeadFirst.Template.Barista/BeverageTestDrive.vb
Imports Microsoft.VisualBasic
Imports System

Namespace DoFactory.HeadFirst.Template.Barista
	Friend Class BeverageTestDrive
        Shared Sub Main(ByVal args() As String)

            Console.WriteLine(Constants.vbLf & "Making tea...")
            Dim tea As New Tea()
            tea.PrepareRecipe()

            Console.WriteLine(Constants.vbLf & "Making coffee...")
            Dim coffee As New Coffee()
            coffee.PrepareRecipe()

            ' Hooked on Template (page 292)

            Console.WriteLine(Constants.vbLf & "Making tea...")
            Dim teaHook As New TeaWithHook()
            teaHook.PrepareRecipe()

            Console.WriteLine(Constants.vbLf & "Making coffee...")
            Dim coffeeHook As New CoffeeWithHook()
            coffeeHook.PrepareRecipe()

            ' Wait for user
            Console.ReadKey()
        End Sub
	End Class

	#Region "Coffee and Tea"

	Public MustInherit Class CaffeineBeverage
		Public Sub PrepareRecipe()
			BoilWater()
			Brew()
			PourInCup()
			AddCondiments()
		End Sub

		Public MustOverride Sub Brew()

		Public MustOverride Sub AddCondiments()

		Private Sub BoilWater()
			Console.WriteLine("Boiling water")
		End Sub

		Private Sub PourInCup()
			Console.WriteLine("Pouring into cup")
		End Sub
	End Class

	Public Class Coffee
		Inherits CaffeineBeverage
		Public Overrides Sub Brew()
			Console.WriteLine("Dripping Coffee through filter")
		End Sub
		Public Overrides Sub AddCondiments()
			Console.WriteLine("Adding Sugar and Milk")
		End Sub
	End Class

	Public Class Tea
		Inherits CaffeineBeverage
		Public Overrides Sub Brew()
			Console.WriteLine("Steeping the tea")
		End Sub
		Public Overrides Sub AddCondiments()
			Console.WriteLine("Adding Lemon")
		End Sub
	End Class

	#End Region

	#Region "Coffee and Tea with Hook"

	Public MustInherit Class CaffeineBeverageWithHook
		Public Sub PrepareRecipe()
			BoilWater()
			Brew()
			PourInCup()
			If CustomerWantsCondiments() Then
				AddCondiments()
			End If
		End Sub

		Public MustOverride Sub Brew()

		Public MustOverride Sub AddCondiments()

		Public Sub BoilWater()
			Console.WriteLine("Boiling water")
		End Sub

		Public Sub PourInCup()
			Console.WriteLine("Pouring into cup")
		End Sub

		Public Overridable Function CustomerWantsCondiments() As Boolean
			Return True
		End Function
	End Class

	Public Class CoffeeWithHook
		Inherits CaffeineBeverageWithHook
		Public Overrides Sub Brew()
			Console.WriteLine("Dripping Coffee through filter")
		End Sub

		Public Overrides Sub AddCondiments()
			Console.WriteLine("Adding Sugar and Milk")
		End Sub

		Public Overrides Function CustomerWantsCondiments() As Boolean
			Dim answer As String = GetUserInput()

			If answer.ToLower().StartsWith("y") Then
				Return True
			Else
				Return False
			End If
		End Function

		Public Function GetUserInput() As String
			Dim answer As String = Nothing
			Console.WriteLine("Would you like milk and sugar with your coffee (y/n)? ")

			Try
				answer = Console.ReadLine()
			Catch
				Console.WriteLine("IO error trying to read your answer")
			End Try

			If answer Is Nothing Then
				Return "no"
			End If
			Return answer
		End Function
	End Class

	Public Class TeaWithHook
		Inherits CaffeineBeverageWithHook

		Public Overrides Sub Brew()
			Console.WriteLine("Steeping the tea")
		End Sub

		Public Overrides Sub AddCondiments()
			Console.WriteLine("Adding Lemon")
		End Sub

		Public Overrides Function CustomerWantsCondiments() As Boolean
			Dim answer As String = GetUserInput()

			If answer.ToLower().StartsWith("y") Then
				Return True
			Else
				Return False
			End If
		End Function

		Private Function GetUserInput() As String
			' get the user's response
			Dim answer As String = Nothing

			Console.WriteLine("Would you like lemon with your tea (y/n)? ")

			Try
				answer = Console.ReadLine()
			Catch
				Console.WriteLine("IO error trying to read your answer")
			End Try

			If answer Is Nothing Then
				Return "no"
			End If
			Return answer
		End Function
	End Class
	#End Region
End Namespace

File: CSharp/HeadFirstPatterns/DoFactory.HeadFirst/DoFactory.HeadFirst.Template.Barista/BeverageTestDrive.cs
using System;

namespace DoFactory.HeadFirst.Template.Barista
{
    class BeverageTestDrive
    {
        static void Main(string[] args)
        {
            Console.WriteLine("\nMaking tea...");
            Tea tea = new Tea();
            tea.PrepareRecipe();

            Console.WriteLine("\nMaking coffee...");
            Coffee coffee = new Coffee();
            coffee.PrepareRecipe();

            // Hooked on Template (page 292)

            Console.WriteLine("\nMaking tea...");
            TeaWithHook teaHook = new TeaWithHook();
            teaHook.PrepareRecipe();

            Console.WriteLine("\nMaking coffee...");
            CoffeeWithHook coffeeHook = new CoffeeWithHook();
            coffeeHook.PrepareRecipe();

            // Wait for user
            Console.ReadKey();
        }
    }

    #region Coffee and Tea

    public abstract class CaffeineBeverage
    {
        public void PrepareRecipe()
        {
            BoilWater();
            Brew();
            PourInCup();
            AddCondiments();
        }

        public abstract void Brew();

        public abstract void AddCondiments();

        void BoilWater()
        {
            Console.WriteLine("Boiling water");
        }

        void PourInCup()
        {
            Console.WriteLine("Pouring into cup");
        }
    }

    public class Coffee : CaffeineBeverage
    {
        public override void Brew()
        {
            Console.WriteLine("Dripping Coffee through filter");
        }
        public override void AddCondiments()
        {
            Console.WriteLine("Adding Sugar and Milk");
        }
    }

    public class Tea : CaffeineBeverage
    {
        public override void Brew()
        {
            Console.WriteLine("Steeping the tea");
        }
        public override void AddCondiments()
        {
            Console.WriteLine("Adding Lemon");
        }
    }

    #endregion

    #region Coffee and Tea with Hook

    public abstract class CaffeineBeverageWithHook
    {
        public void PrepareRecipe()
        {
            BoilWater();
            Brew();
            PourInCup();
            if (CustomerWantsCondiments())
            {
                AddCondiments();
            }
        }

        public abstract void Brew();

        public abstract void AddCondiments();

        public void BoilWater()
        {
            Console.WriteLine("Boiling water");
        }

        public void PourInCup()
        {
            Console.WriteLine("Pouring into cup");
        }

        public virtual bool CustomerWantsCondiments()
        {
            return true;
        }
    }

    public class CoffeeWithHook : CaffeineBeverageWithHook
    {
        public override void Brew()
        {
            Console.WriteLine("Dripping Coffee through filter");
        }

        public override void AddCondiments()
        {
            Console.WriteLine("Adding Sugar and Milk");
        }

        public override bool CustomerWantsCondiments()
        {
            string answer = GetUserInput();

            if (answer.ToLower().StartsWith("y"))
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        public string GetUserInput()
        {
            string answer = null;
            Console.WriteLine("Would you like milk and sugar with your coffee (y/n)? ");

            try
            {
                answer = Console.ReadLine();
            }
            catch
            {
                Console.WriteLine("IO error trying to read your answer");
            }

            if (answer == null)
            {
                return "no";
            }
            return answer;
        }
    }

    public class TeaWithHook : CaffeineBeverageWithHook
    {

        public override void Brew()
        {
            Console.WriteLine("Steeping the tea");
        }

        public override void AddCondiments()
        {
            Console.WriteLine("Adding Lemon");
        }

        public override bool CustomerWantsCondiments()
        {
            string answer = GetUserInput();

            if (answer.ToLower().StartsWith("y"))
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        private string GetUserInput()
        {
            // get the user's response
            string answer = null;

            Console.WriteLine("Would you like lemon with your tea (y/n)? ");

            try
            {
                answer = Console.ReadLine();
            }
            catch
            {
                Console.WriteLine("IO error trying to read your answer");
            }

            if (answer == null)
            {
                return "no";
            }
            return answer;
        }
    }
    #endregion
}
